src/replication.c

```diff
@@ -2071,14 +2451,180 @@ handle_update_spec_stats(spec_t *spec, zvol_io_hdr_t *hdr, void *resp)
 	return;
 }
 
+/*
+ * wait_for_ongoing_ios_on_replica pause the IOs and it's caller
+ * responsibility to pause IO's and call this function
+ * Returns:
+ * 0 - on success
+ * -1 - still, if pending IOs are there
+ */
+static int
+wait_for_ongoing_ios_on_replica(spec_t *spec, replica_t *replica, int sec) {
+	int ret = -1;
+	struct timespec last, now, diff;
+	bool io_found = false;	/* Write or Sync IOs */
+
+	ASSERT(MTX_LOCKED(&spec->rq_mtx));
+	ASSERT(spec->quiesce == 1);
+
+	clock_gettime(CLOCK_MONOTONIC_COARSE, &last);
+	timesdiff(CLOCK_MONOTONIC_COARSE, last, now, diff);
+
+	while (diff.tv_sec < sec) {
+		io_found = false;
+
+		/*
+		 * Check on spec is required as these IOs as well
+		 * need to be pushed to replica before setting scalingup_replica
+		 * to replica.
+		 */
+		if (spec->inflight_write_io_cnt != 0 ||
+		    spec->inflight_sync_io_cnt != 0)
+			io_found = true;
+		if (replica->replica_inflight_write_io_cnt != 0 ||
+		    replica->replica_inflight_sync_io_cnt != 0) {
+			io_found = true;
+		}
+		if (!io_found) {
+			ret = 0;
+			break;
+		}
+
+		MTX_UNLOCK(&spec->rq_mtx);
+
+		ISTGT_LOG("Waiting for IO's to flush on replica(%s:%lu) "
+		    "spec write_io_cnt: %lu and sync_io_cnt: %lu replica "
+		    "write_io_cnt: %lu sync_io_cnt: %lu\n",
+	   	    replica->replica_id, replica->zvol_guid,
+		    spec->inflight_write_io_cnt, spec->inflight_sync_io_cnt,
+		    replica->replica_inflight_write_io_cnt, replica->replica_inflight_sync_io_cnt);
+		/*
+		 * inflight write/sync IOs in spec, or in replica,
+		 * so, wait for some time
+		 */
+		sleep (1);
+		MTX_LOCK(&spec->rq_mtx);
+		timesdiff(CLOCK_MONOTONIC_COARSE, last, now, diff);
+	}
+
+	return ret;
+}
+
+//TODO: Need to restructure handle_scaleup_replica_transition and
+// handle_transition_from_unknown_to_known
+
+/* handle_scaleup_replica_transition should be called only when
+ * there is replica scaleup case
+ * Returns
+ * = 0 if success
+ * < 0 for caller to retry
+ */
+static int
+handle_scaleup_replica_transition(spec_t *spec, replica_t *replica)
+{
+	int rf, cf, rc;
+	int pause_io_wait_time = 10;
+	/*
+	 * Check if there is no change in CF wrt new RF, i.e., (RF + 1)
+	 */
+	rf = spec->replication_factor + 1;
+	cf = (rf/2) + 1;
+	if (spec->consistency_factor == cf)
+		rc = update_trusty_replica_list(spec, replica, rf);
+	else {
+
+		rc = 0;
+
+		/*
+		 * Wait for ongoing IOs if first time (or)
+		 * waiting in previous iteration was NOT successful.
+		 */
+
+		if (spec->scalingup_replica == NULL) {
+			/* Pause IO's */
+			spec->quiesce = 1;
+
+			rc = wait_for_ongoing_ios_on_replica(spec, replica, pause_io_wait_time);
+
+			/* set transition replica so that new consistency model as well gets applied */
+			if (rc == 0)
+				spec->scalingup_replica = replica;
+
+			/* Resume IOs */
+			spec->quiesce = 0;
+		}
+
+		/* Retry again */
+		if (rc != 0)
+			return rc;
+
+		rc = update_trusty_replica_list(spec, replica, rf);
```

Comment:
can 'rf' be set in update in_memory replica list itself?

Reply:
I think if we do this then istgt(target) is tightly coupled with control plane because we are setting cf as `(rf+1)/2+1`. If that is fine then we can move it into that